tests/uniswap_crafter.py

```diff
@@ -0,0 +1,150 @@
+import json
+import os
+import string
+
+import datetime
+
+from typing import Union, List, Optional
+
+class TokenAndNetworkFee:
+    address: str
+    network_fee: Optional[str]
+
+    def __init__(self, address, network_fee=None):
+        self.address = address
+        self.network_fee = network_fee
+
+def _address_to_parameter(address: str) -> str:
+    return '00' * 12 + address
+
+def _integer_to_parameter(var: int) -> str:
+    hex_string = hex(var)[2:]  # Remove the '0x' prefix
+    padded_hex_string = '0' * (64 - len(hex_string)) + hex_string
+    return padded_hex_string
+
+def _bool_to_parameter(var: bool) -> str:
+    if var:
+        return _integer_to_parameter(1)
+    else:
+        return _integer_to_parameter(0)
+
+def _craft_v2_path(in_token: str, intermediate_tokens: [str], out_token: str) -> str:
+    path = ""
+    path_length = 1 + len(intermediate_tokens) + 1
+    path += _integer_to_parameter(path_length)
+    path += _address_to_parameter(in_token)
+    for step in intermediate_tokens:
+        path += _address_to_parameter(step)
+    path += _address_to_parameter(out_token)
+    return path
+
+def _craft_v3_path(in_token: str, intermediate_tokens: [TokenAndNetworkFee], out_token: TokenAndNetworkFee) -> str:
+    path = ""
+    path_length = 20 + (len(intermediate_tokens) * (3 + 20)) + (3 + 20)
+    path += _integer_to_parameter(path_length)
+    path += in_token
+    for step in intermediate_tokens:
+        path += step.network_fee
+        path += step.address
+    path += out_token.network_fee
+    path += out_token.address
+    padding = 32 - (path_length % 32)
+    path += '00' * padding
+    return path
+
+def _craft_v3_path_reversed(in_token: str, intermediate_tokens: [TokenAndNetworkFee], out_token: TokenAndNetworkFee) -> str:
+    path = ""
+    path_length = (20 + 3) + (len(intermediate_tokens) * (20 + 3)) + 20
+    path += _integer_to_parameter(path_length)
+    path += out_token.address
+    path += out_token.network_fee
+    for step in intermediate_tokens:
+        path += step.address
+        path += step.network_fee
+    path += in_token
+    padding = 32 - (path_length % 32)
+    path += '00' * padding
+    return path
+
+class UniswapCommand:
+    command_byte: bytes
+    input: bytes
+
+    def __init__(self, command_byte, input):
+        self.command_byte = command_byte
+        self.input = input
+
+def craft_V3_SWAP_EXACT_IN(in_token: str, intermediate_tokens: [TokenAndNetworkFee], out_token: TokenAndNetworkFee, amount_in: int, amount_out: int, recipient: str = "0000000000000000000000000000000000000001") -> UniswapCommand:
+    input_as_str = "000000000000000000000000" + recipient
```

Comment:
All this manual calldata crafting could be simplified by using something like : https://eth-abi.readthedocs.io/en/latest/encoding.html#encoding-python-values

Reply:
Done